### PR TITLE
Add Cadl doc for the OpenAPI developer

### PR DESCRIPTION
**Pull request by mikekistler**
This PR creates an initial version of "Cadl for the OpenAPI developer".  There are no doubt areas where it could be improved or expanded, but I think the basic structure and initial content are sufficient to help developers that know OpenAPI at some level to quickly get started with Cadl.

### CODE REVIEW

File: docs/cadl-for-openapi-dev.md [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+## Consumes / Produces (OAS2)
+
+In OpenAPI v2, the top-level `consumes` and `produces` fields specify a list of MIME types an operation can consume / produce
+when not overridden by a `consumes` or `produces` on an individual operation.
+
+In Cadl, these global values for `consumes` and `produces` are defined with the `@consumes` and `@produces` decorators.
+
+```cadl
+@produces("application/json", "image/png")
+@consumes("application/json", "application/octet-stream")
```
Comment (timotheeguerin) [inline docs/cadl-for-openapi-dev.md:88]: This needs to be applied on the global namespace right?

File: docs/cadl-for-openapi-dev.md [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+namespace Pets {
+  op read(... PetId): Pet | Error;
+}
+
+```
+
+results in a `$ref` to the named parameter `PetId` in either `parameters` or `components.parameters`.
+
+## Security Schemes Object
```
Comment (timotheeguerin) [inline docs/cadl-for-openapi-dev.md:272]: missing not supported in this section maybe?

File: docs/cadl-for-openapi-dev.md [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+OpenAPI schemas are represented in Cadl by [models](https://github.com/microsoft/cadl/blob/main/docs/tutorial.md#models).
+Models have any number of members and can extend and be composed with other models.
+
+Models can be defined with the `model` statement and then referenced by name, which generally results in a `$ref` to a schema for the model in the `definitions` or `components.schemas` section of the OpenAPI document.
+
+Cadl supports the ["spread" operator](https://github.com/microsoft/cadl/blob/main/docs/tutorial.md#spread) (`...`), which copies the members of the source model into the target model.
+But Cadl processes all spread transformations before emitters are invoked, so this form of reuse is not represented in the emitted OpenAPI.
+
+Cadl also supports single inheritance of models with the `extends` keyword. This construct can be used to produce an `allOf` with a single element (the parent schema) in OpenAPI. Cadl does not current provide a means to produce an `allOf` with more than one element -- these are generally treated as "composition" in code generators and thus better represented in Cadl with the spread operator.
```
Comment (timotheeguerin) [inline docs/cadl-for-openapi-dev.md:247]: Can we have an example of `extends` and `....` so they know where to use it without having to look at the main tutorial

File: docs/cadl-for-openapi-dev.md [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+| --------------------------- | ----------------------------- | ----- |
+| `minLength: value`          | `@minLength(value)` decorator |       |
+| `maxLength: value`          | `@maxLength(value)` decorator |       |
+| `pattern: regex`            | `@format(regex)` decorator    |       |
+
+To define a `type: string` parameter or property with an `enum`, define the Cadl element as the union of the enum values.
+For example:
+
+```cadl
+status: "Running" | "Stopped" | "Failed"
```
Comment (timotheeguerin) [inline docs/cadl-for-openapi-dev.md:57]: might want to describe the actual `enum` keyword too and when you should use which

File: docs/cadl-for-openapi-dev.md [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+| OpenAPI `type`/`format`           | Cadl type        | Notes                  |
+| --------------------------------- | ---------------- | ---------------------- |
+| `type: integer, format: int32`    | `int32`          |                        |
+| `type: integer, format: int64`    | `int64`          |                        |
+| `type: number, format: float`     | `float32`        |                        |
+| `type: number, format: double`    | `float64`        |                        |
+| `type: string`                    | `string`         |                        |
+| `type: string, format: byte`      |                  |                        |
+| `type: string, format: binary`    | `byte[]`         | any sequence of octets |
```
Comment (markcowl) [inline docs/cadl-for-openapi-dev.md:29]: The ```byte[]``` type should correspond to ```type: string, format: byte```, it is meant to be interpreted as a base64-encoded string.  We don't yet have a representation for a stream of octets.

File: docs/cadl-for-openapi-dev.md (hunk 4 above)
Comment (markcowl) [reply docs/cadl-for-openapi-dev.md:57]: Agreed, enum would be a better representation than union of string

File: docs/cadl-for-openapi-dev.md [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+[v2-info]: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#infoObject
+[v3-info]: https://github.com/OAI/OpenAPI-Specification/blob/3.0.3/versions/3.0.3.md#infoObject
+
+In Cadl this information is specified with [decorators on the namespace][cadl-service-metadata].
+
+| OpenAPI `info` field | Cadl decorator    | Notes                    |
+| -------------------- | ----------------- | ------------------------ |
+| `title`              | `@serviceTitle`   |                          |
+| `version`            | `@serviceVersion` |                          |
```
Comment (markcowl) [inline docs/cadl-for-openapi-dev.md:72]: Should include the host as part of this metadata: https://github.com/microsoft/cadl/blob/c4c517084f998a3171ed369e42b79941c67f5652/packages/compiler/lib/service.ts#L71-L81

File: docs/cadl-for-openapi-dev.md [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+| `description`        |                   | Not currently supported. |
```
Comment (markcowl) [inline docs/cadl-for-openapi-dev.md:73]: I believe that an `@doc` added to the default namespace should be the description

**Review by markcowl — CHANGES_REQUESTED** [review]
Mainly, this looks very good.  I am a bit worried about the impression that CADL is meant to be a 1:1 mapping for Swagger, or that CADL will, at some point, support everything you can describe in swagger, but I think this will be really valuable to folks familiar with swagger in getting started.

File: docs/cadl-for-openapi-dev.md [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+| `pattern: regex`            | `@format(regex)` decorator    |       |
+
+To define a `type: string` parameter or property with an `enum`, define the Cadl element as the union of the enum values.
+For example:
+
+```cadl
+status: "Running" | "Stopped" | "Failed"
+```
+
+## Info Object
```
Comment (markcowl) [inline docs/cadl-for-openapi-dev.md:60]: I wonder if this should go further down in the doc?  Filling out the info object is not generally a primary concern when getting started.

File: docs/cadl-for-openapi-dev.md [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+| -------------------- | ----------------- | ------------------------ |
+| `title`              | `@serviceTitle`   |                          |
+| `version`            | `@serviceVersion` |                          |
+| `description`        |                   | Not currently supported. |
+| `license`            |                   | Not currently supported. |
+| `contact`            |                   | Not currently supported. |
+
+[cadl-service-metadata]: https://github.com/microsoft/cadl/blob/main/docs/tutorial.md#service-definition-and-metadata
+
+## Consumes / Produces (OAS2)
```
Comment (markcowl) [inline docs/cadl-for-openapi-dev.md:79]: I think this should go further down.  As it is, consumes/produces can be specified, but can also be inferred from request and response data

**Review by markcowl — CHANGES_REQUESTED** [review]

File: docs/cadl-for-openapi-dev.md (hunk 1 above)
Comment (mikekistler) [reply docs/cadl-for-openapi-dev.md:88]: Maybe, but the tutorial doesn't say anything about a "global" namespace. Does this need to be described in the tutorial, or is there some other word for it in the tutorial?  Or can we say "outer-most"? Could there be more than one "outer-most" namespace?  If so, what then?

File: docs/cadl-for-openapi-dev.md (hunk 2 above)
Comment (mikekistler) [reply docs/cadl-for-openapi-dev.md:272]: These can be set through JavaScript so I added that.

File: docs/cadl-for-openapi-dev.md (hunk 8 above)
Comment (mikekistler) [reply docs/cadl-for-openapi-dev.md:60]: Fair point. I was mostly following the order of the elements in the OpenAPI spec, but I suppose we can deviate from that where appropriate.

File: docs/cadl-for-openapi-dev.md (hunk 5 above)
Comment (bterlson) [reply docs/cadl-for-openapi-dev.md:29]: Also, I think I replaced byte with bytes some time ago?

File: docs/cadl-for-openapi-dev.md (hunk 1 above)
Comment (bterlson) [reply docs/cadl-for-openapi-dev.md:88]: IIRC they go on the namespace bearing the service metadata, presumably an emitter could support multiple namespaces bearing this metadata via something like walking up from an operation to the nearest namespace containing the metadata (and emitting multiple  json files as necessary), but I suspect that doesn't work.

File: docs/cadl-for-openapi-dev.md (hunk 7 above)
Comment (mikekistler) [reply docs/cadl-for-openapi-dev.md:73]: It should be but it is not.

File: docs/cadl-for-openapi-dev.md (hunk 9 above)
Comment (mikekistler) [reply docs/cadl-for-openapi-dev.md:79]: Moved.  But I think the logic that does this inference has some bugs. I'll open an issue.

**Comment by mikekistler** [comment]
Thanks @timotheeguerin @markcowl @bterlson for your comments.  I just pushed a new commit that I think addresses all of these.

**Review by markcowl — APPROVED** [review]

**Pull request merged**